app/src/main/java/ca/marshallasch/veil/DataStore.java

```diff
@@ -0,0 +1,198 @@
+package ca.marshallasch.veil;
+
+import android.content.Context;
+import android.support.annotation.Nullable;
+import android.support.v4.util.Pair;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import ca.marshallasch.veil.comparators.CommentPairComparator;
+import ca.marshallasch.veil.database.Database;
+import ca.marshallasch.veil.exceptions.TooManyResultsException;
+import ca.marshallasch.veil.proto.DhtProto;
+import ca.marshallasch.veil.proto.Sync;
+
+/**
+ * @author Marshall Asch
+ * @version 1.0
+ * @since 2018-06-15
+ */
+public class DataStore
+{
+    private Database db;
+    private HashTableStore hashTableStore;
+
+    private static DataStore instance;
+    private static final AtomicInteger openCounter = new AtomicInteger();
+
+    private DataStore(Context context) {
+
+        db = Database.getInstance(context);
+        hashTableStore = HashTableStore.getInstance(context);
+
+    }
+
+    public static synchronized DataStore getInstance(Context context){
+
+        if (instance == null) {
+            instance = new DataStore(context);
+        }
+
+        openCounter.getAndIncrement();
+        return instance;
+    }
+
+    /**
+     * Save the hash table to a persistent file.
+     * @param context the application context required to save a file.
+     */
+    public void save(Context context) {
+        hashTableStore.save(context);
+    }
+
+    /**
+     * Save a post to the hash table and add its key to the database.
+     * @param post the post to insert
+     * @return true on success false oon failure.
+     */
+    public boolean savePost(@Nullable DhtProto.Post post) {
+
+        if (post == null) {
+            return false;
+        }
+
+        String postHash = hashTableStore.insertPost(post);
+
+        // insert any necessary query keys in here.
+
+        return db.insertKnownPost(postHash, null);
+    }
+
+    /**
+     * Gets all the known posts in a the data store
+     * @return the list of posts pairs.
+     */
+    public List<Pair<String, DhtProto.Post>> getKnownPosts() {
+
+        List<String> hashes = db.getPostHashes();
+
+        Pair<String, DhtProto.Post> postPair;
+        List<Pair<String, DhtProto.Post>> posts = new ArrayList<>();
```

Comment:
Is it necessary to store as a pair here? I'm assuming the String param of the <String, Dht.Proto.Post> is referring to the hash of the post. Which I don't feel is necessary to include since I cant imagine where else we would need it after receiving it.

Reply:
Fixed in commit 6d83384.

**NOTE:** that the change to move the hash into the body of the post object will cause any old data that is in the hash table to no longer work correctly.